Onboarding Users to BioHub

dwhogan opened the pull request:
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-838

## Description of Changes

As a BioHub user, I want to be added as a system user when I first sign in, so that my account is properly created and recognized by the system.

Acceptance Criteria

With a PUT to /api/user/self

1. Check for Existing System User  
WHEN a user logs into BioHub,  
  THEN the system checks whether a matching record exists in the `system_user` table,  
  AND the match should be based on the user's `Keycloak GUID` where `record_end_date` is `NULL`.
See useBiohubUserWrapper.tsx and useAuthStateContext.tsx

2. Existing User Found  
WHEN a matching active system user record exists,  
  THEN the user is allowed to proceed with login without creating a new record. display_name and email are update if required
  AND return 200 with user 


3. No Existing User Found  
WHEN no active system user record is found for the user's Keycloak GUID,  
  THEN the system should allow login,  
  AND should create a new system user record
  AND return 201 with user

4. Invalid or Expired User Record  
WHEN a system user record exists but is invalid or expired (e.g., `record_end_date` is in the past),  
THEN return 401 unauthorized,

5. Required User Fields  
WHEN creating a new system user,  
  THEN the record must include:  
  • `user_identity_source`  
  • `user_identifier`  
  • `user_guid`  
AND a new record should be inserted into `system_user_role` with the 'Member' role
NOTE: Biohub added the following fields after the codebases diverged:
  ADD COLUMN display_name varchar(100)
  ADD COLUMN given_name varchar(100)
  ADD COLUMN family_name varchar(100)
  ADD COLUMN email varchar(100)
  ADD COLUMN agency varchar(100)
  ADD COLUMN notes varchar(250)


6. Auditing and Integrity  
There should be a unique constraint on keycloak guid in the `system_user` table

## Testing Notes
 
  • Remove you self from the sys user seed file

  • Open web page http://localhost:7100/search

  • Check if new user and role records are created

  • Change user's email in system user table

 • Open web page http://localhost:7100/search, check if email is updated in the system users table

 • Deactivated user in the system user table

 • Open web page http://localhost:7100/search check console for 401 status on self endpoint

mauberti-bc commented on app/src/hooks/api/useUserApi.ts:
```diff
@@ -31,6 +32,48 @@ const useUserApi = (axios: AxiosInstance) => {
+   *
+   * @return {*}  {Promise<ISystemUser>}
+   */
+  const getOrRegisterUser = async (): Promise<ISystemUser> => {
```
This logic can be moved to the backend, to the /self endpoint. 
* If the user exists and is valid -> return the user
* If the user exists but is invalid -> return 401 unauthorized
* If the user does not exist -> create the user

If a user makes a request to the /self endpoint, it should return 200 with the user record unless the user has an existing invalid record in the system_user table, in which case it returns 401 unauthorized.

mauberti-bc commented on api/src/paths/user/self.ts:
```diff
@@ -163,3 +94,101 @@ export function getUser(): RequestHandler {
+ *
+ * @returns {RequestHandler}
+ */
+export function registerUser(): RequestHandler {
```
I agree with your comments about making this more like an upsert. See the comment below about handling the 3 scenarios (user exists and is valid, user exists and is invalid, and user doesn't exist).

dwhogan committed:
feat: implement user auto-registration with profile fields

- Add PUT /api/user/self endpoint for user upsert
- Add user profile fields (display_name, given_name, family_name, email)
- Add database migrations for profile fields and user GUID uniqueness
- Simplify frontend to use upsert endpoint

dwhogan committed:
fix(sonar): Async method 'addSystemUser' has too many parameters

mauberti-bc reviewed (changes requested):
I tested this by removing myself from the DB seeds, then logging in as a new user. I'm getting a Zod error from db-constants.ts -- you might have the ```DATABASE_RESPONSE_VALIDATION_ENABLED``` set to false? It should be true.

The issue is that the SystemUser zod object requires the new fields, but ```const selectServiceAccountsSqlStatement = ...``` on line 63 of db-constants.ts doesn't select them.

dwhogan committed:
Merge branch 'dev' into SIMSBIOHUB-838

dwhogan committed:
fix(app): Reviewer fix selectServiceAccountsSqlStatement

dwhogan committed:
fix(database): add dahogan to system users seed

mauberti-bc commented on api/src/paths/user/self.ts:
```diff
@@ -19,93 +30,21 @@ GET.apiDoc = {
       content: {
         'application/json': {
           schema: {
-            title: 'User Response Object',
```
I think the system user OpenAPI schema defined in ```root-api-doc.ts``` is just missing the new fields (family_name, given_name, etc.)

<img width="1195" height="764" alt="image" src="https://github.com/user-attachments/assets/11f6e7cc-ba75-4bb9-92fd-7cd3533d73a4" />

mauberti-bc reviewed (changes requested):
Looks great, other than the error!

dwhogan committed:
fix(api): add profile properties to SystemUser schema

mauberti-bc reviewed (approved):
👍

dwhogan committed:
Merge branch 'dev' into SIMSBIOHUB-838
